crates/lsp/src/logging.rs

```diff
@@ -0,0 +1,265 @@
+        let contents = contents.to_string();
+
+        // Forward the log message to the latency sensitive log thread,
+        // which is in charge of forwarding to the client in an async manner.
+        self.log_tx
+            .send(LogMessage { contents })
+            .map_err(|e| IoError::new(ErrorKind::Other, e))?;
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+struct LogWriterMaker {
+    log_tx: LogMessageSender,
+}
+
+impl LogWriterMaker {
+    fn new(log_tx: LogMessageSender) -> Self {
+        Self { log_tx }
+    }
+}
+
+impl<'a> MakeWriter<'a> for LogWriterMaker {
+    type Writer = LogWriter<'a>;
+
+    fn make_writer(&'a self) -> Self::Writer {
+        LogWriter::new(&self.log_tx)
+    }
+}
+
+pub(crate) fn init_logging(
+    log_tx: LogMessageSender,
+    log_level: Option<LogLevel>,
+    client_info: Option<&ClientInfo>,
+) {
+    let log_level = resolve_log_level(log_level);
+
+    let writer = if client_info.is_some_and(|client_info| {
+        client_info.name.starts_with("Zed") || client_info.name.starts_with("Visual Studio Code")
+    }) {
+        BoxMakeWriter::new(LogWriterMaker::new(log_tx))
+    } else {
+        BoxMakeWriter::new(std::io::stderr)
+    };
+
+    let layer = tracing_subscriber::fmt::layer()
+        // Disable ANSI escapes, those are not supported in Code
+        .with_ansi(false)
+        // Display source code file paths
+        .with_file(true)
+        // Display source code line numbers
+        .with_line_number(true)
+        // Don't display the thread ID or thread name
+        .with_thread_ids(false)
+        .with_thread_names(false)
+        // Don't display the event's target (module path).
+        // Mostly redundant with file paths.
+        .with_target(false)
+        // Display local time rather than UTC
+        .with_timer(LocalTime::rfc_3339())
+        // Display the log level
+        .with_level(true)
+        .with_writer(writer)
+        .with_filter(LogLevelFilter { filter: log_level });
+
+    let subscriber = tracing_subscriber::Registry::default().with(layer);
+
+    if !is_test_client(client_info) {
+        tracing::subscriber::set_global_default(subscriber)
+            .expect("Should be able to set the global subscriber.");
+    }
+
+    tracing::info!("Logging initialized with level: {log_level}");
+}
+
+/// We never log during tests as tests run in parallel within a single process,
+/// but you can only have 1 global subscriber per process.
+///
+/// If you are debugging a single test, you can override this to emit messages to stderr.
+///
+/// Note that if you override this and run multiple tests in parallel, then the call
+/// to `set_global_default()` will error causing a panic.
+fn is_test_client(client_info: Option<&ClientInfo>) -> bool {
+    client_info.map_or(false, |client_info| client_info.name == "AirTestClient")
+}
```

Comment:
Important notes here about skipping logging during tests. They are very noisy in the test output, and there are issues with running tests in parallel since the "global subscriber" is by nature a "once per process" kind of thing.

Reply:
That seems strange to me. It feels like it would be better to control this with the log level. By default we don't see this output since it's captured anyway.

Tests are run in parallel that's true, but I don't think it's customary to disable logging in tests altogether? It seems like we're making it harder to debug tests here.